File: Benchmarks/Core/Scene/Scene.cs
using Benchmarks.Core.Entity;
using Benchmarks.Core.Systems;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Scene
{
    /// <summary>
    /// La representacion de una escena del juego.
    /// Replica exacta del lifecycle del engine: Initialize → Update → Draw → Dispose.
    /// </summary>
    public class Scene : IDisposable
    {
        public string SceneName { get; set; }
        public EntityManager EntityManager { get; set; }
        public EventManager EventManager { get; set; }
        public SystemBuilder SystemBuilder { get; set; }

        internal FontSystem2D FontSystem2D;
        private AnimationSystem2D AnimationSystem2D;
        private ButtonSystem2D ButtonSystem2D;
        private PhysicsSystem2D PhysicsSystem2D;
        private RenderSystem2D RenderSystem2D;
        private CameraSystem CameraSystem;
        private InputSystem InputSystem;
        public ScriptSystem ScriptSystem;
        private TileMapSystem2D TilemapSystem;

        public Scene()
        {
            EventManager = EventManager.Instance;
            EntityManager = new();
            AnimationSystem2D = new AnimationSystem2D();
            ButtonSystem2D = new ButtonSystem2D();
            PhysicsSystem2D = new PhysicsSystem2D();
            RenderSystem2D = new RenderSystem2D();
            CameraSystem = new CameraSystem();
            InputSystem = new InputSystem();
            ScriptSystem = new();
            TilemapSystem = new();
            FontSystem2D = new();
            SystemBuilder = new SystemBuilder();
        }

        /// <summary>
        /// Primer metodo que se ejecuta y prepara todos los sistemas.
        /// </summary>
        public void Initialize()
        {
            AnimationSystem2D.InitializeSystem(EntityManager);
            ButtonSystem2D.InitializeSystem(EntityManager);
            PhysicsSystem2D.InitializeSystem(EntityManager);
            RenderSystem2D.InitializeSystem(EntityManager);
            CameraSystem.InitializeSystem(EntityManager);
            InputSystem.InitializeSystem(EntityManager);
            ScriptSystem.InitializeSystem(EntityManager);
            TilemapSystem.InitializeSystem(EntityManager);
            FontSystem2D.InitializeSystem(EntityManager);
            SystemBuilder.InitializeSystem(EntityManager);

            foreach (ref Yotsuba entity in EntityManager.YotsubaEntities.AsSpan())
            {
                SystemBuilder.SharedEntityInitialize(ref entity);
                ScriptSystem.SharedEntityInitialize(ref entity);
                FontSystem2D.SharedEntityInitialize(ref entity);
            }
        }

        /// <summary>
        /// Metodo que se ejecuta en cada frame para actualizar el estado del juego.
        /// </summary>
        public void Update(GameTime gameTime)
        {
            InputSystem.UpdateSystem(gameTime);

            if (EntityManager.YotsubaEntities == null || EntityManager.YotsubaEntities.Count == 0) return;
            PhysicsSystem2D.UpdateSystem(gameTime);
            ButtonSystem2D.UpdateSystem(gameTime);
            AnimationSystem2D.UpdateSystem(gameTime);
            CameraSystem.UpdateSystem(gameTime);
            SystemBuilder.UpdateSystem(gameTime);

            foreach (ref Yotsuba entity in EntityManager.YotsubaEntities.AsSpan())
            {
                ScriptSystem.SharedEntityForEachUpdate(ref entity, gameTime);
                SystemBuilder.SharedEntityForEachUpdate(ref entity, gameTime);
            }
        }

        /// <summary>
        /// Metodo que se ejecuta en cada frame para dibujar la escena.
        /// </summary>
        public void Draw(GameTime gameTime)
        {
            if (EntityManager.YotsubaEntities != null && EntityManager.YotsubaEntities.Count > 0)
            {
                Rectangle cameraBounds = new Rectangle(0, 0, 1920, 1080);
                TilemapSystem.UpdateSystem(gameTime, cameraBounds);

                RenderSystem2D.UpdateSystem(gameTime);

                FontSystem2D.DrawSystem(gameTime);

                // Draw3D equivalent
                ScriptSystem.DrawSystem3D(gameTime);
                SystemBuilder.Render3D(gameTime);

                // Draw2D equivalent
                ScriptSystem.DrawSystem2D(gameTime);
                SystemBuilder.Render2D(gameTime);
            }
        }

        public void Clear()
        {
            ScriptSystem.Clear();
            SystemBuilder.Dispose();
            AnimationSystem2D.Dispose();
            ButtonSystem2D.Dispose();
            PhysicsSystem2D.Dispose();
            CameraSystem.Dispose();
            InputSystem.Dispose();
            FontSystem2D.Dispose();
        }

        public void Dispose()
        {
            Clear();
        }
    }
}

File: Benchmarks_FSharp/Core/Scene/Scene.fs
namespace Benchmarks.Core.Scene

open System
open Benchmarks.Core.Entity
open Benchmarks.Core.Systems
open Benchmarks.Core.Types

type Scene() =
    member val SceneName: string = null with get, set
    member val EntityManager = EntityManager() with get, set
    member val EventManager = EventManager.Instance with get, set
    member val SystemBuilder = SystemBuilder() with get, set

    member val internal FontSystem2D = FontSystem2D() with get, set
    member val private AnimationSystem2D = AnimationSystem2D() with get, set
    member val private ButtonSystem2D = ButtonSystem2D() with get, set
    member val private PhysicsSystem2D = PhysicsSystem2D() with get, set
    member val private RenderSystem2D = RenderSystem2D() with get, set
    member val private CameraSystem = CameraSystem() with get, set
    member val private InputSystem = InputSystem() with get, set
    member val ScriptSystem = ScriptSystem() with get, set
    member val private TilemapSystem = TileMapSystem2D() with get, set

    member this.Initialize() =
        this.AnimationSystem2D.InitializeSystem(this.EntityManager)
        this.ButtonSystem2D.InitializeSystem(this.EntityManager)
        this.PhysicsSystem2D.InitializeSystem(this.EntityManager)
        this.RenderSystem2D.InitializeSystem(this.EntityManager)
        this.CameraSystem.InitializeSystem(this.EntityManager)
        this.InputSystem.InitializeSystem(this.EntityManager)
        this.ScriptSystem.InitializeSystem(this.EntityManager)
        this.TilemapSystem.InitializeSystem(this.EntityManager)
        this.FontSystem2D.InitializeSystem(this.EntityManager)
        this.SystemBuilder.InitializeSystem(this.EntityManager)

        let entities = this.EntityManager.YotsubaEntities.AsSpan()
        for i = 0 to entities.Length - 1 do
            let entity = &entities.[i]
            this.SystemBuilder.SharedEntityInitialize(&entity)
            this.ScriptSystem.SharedEntityInitialize(&entity)
            this.FontSystem2D.SharedEntityInitialize(&entity)

    member this.Update(gameTime: GameTime) =
        this.InputSystem.UpdateSystem(gameTime)

        if isNull this.EntityManager.YotsubaEntities || this.EntityManager.YotsubaEntities.Count = 0 then ()
        else
        this.PhysicsSystem2D.UpdateSystem(gameTime)
        this.ButtonSystem2D.UpdateSystem(gameTime)
        this.AnimationSystem2D.UpdateSystem(gameTime)
        this.CameraSystem.UpdateSystem(gameTime)
        this.SystemBuilder.UpdateSystem(gameTime)

        let entities = this.EntityManager.YotsubaEntities.AsSpan()
        for i = 0 to entities.Length - 1 do
            let entity = &entities.[i]
            this.ScriptSystem.SharedEntityForEachUpdate(&entity, gameTime)
            this.SystemBuilder.SharedEntityForEachUpdate(&entity, gameTime)

    member this.Draw(gameTime: GameTime) =
        if not (isNull this.EntityManager.YotsubaEntities) && this.EntityManager.YotsubaEntities.Count > 0 then
            let cameraBounds = Rectangle(0, 0, 1920, 1080)
            this.TilemapSystem.UpdateSystem(gameTime, cameraBounds)
            this.RenderSystem2D.UpdateSystem(gameTime)
            this.FontSystem2D.DrawSystem(gameTime)
            this.ScriptSystem.DrawSystem3D(gameTime)
            this.SystemBuilder.Render3D(gameTime)
            this.ScriptSystem.DrawSystem2D(gameTime)
            this.SystemBuilder.Render2D(gameTime)

    member this.Clear() =
        this.ScriptSystem.Clear()
        this.SystemBuilder.Dispose()
        this.AnimationSystem2D.Dispose()
        this.ButtonSystem2D.Dispose()
        this.PhysicsSystem2D.Dispose()
        this.CameraSystem.Dispose()
        this.InputSystem.Dispose()
        this.FontSystem2D.Dispose()

    interface IDisposable with
        member this.Dispose() = this.Clear()
